src/dmd/expressionsem.d

```diff
@@ -4761,6 +4761,48 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
         if (exp.e1.op == TOK.dotVariable)
         {
             DotVarExp dve = cast(DotVarExp)exp.e1;
+
+            // Fix Issue 14245 : https://issues.dlang.org/show_bug.cgi?id=14245
+            // The address of an uninitialized immutable field cannot be taken
+            // inside a constructor because it will break immutability once
+            // the field is assigned
+            if (sc.func)
+            {
+                // if inside a constructor
+                if(const ctd = sc.func.isCtorDeclaration())
+                {
+                    const decl = dve.var;
+
+                    // The below condition checks for (&this.var && var is immutable). Note that
+                    // var might not be declared immutable, but if the address is taken inside an
+                    // immutable constructor then var will be part  of an immutable object, thus
+                    // being immutable. This case can be identified by checking if the constructor
+                    // is immutable.
+                    if (dve.e1.op == TOK.this_ && decl.isField() && (decl.isImmutable() || ctd.type.isImmutable()))
+                    {
+                        const dim = sc.ctorflow.fieldinit.length;
+                        auto ad = ctd.isMember2();
+
+                        size_t i;
+                        // obtain the position of the variable in the fieldinit array
+                        for (i = 0; i < dim; i++)
```

Comment:
Can we add an `indexOf` method to the `dmd.root.Array`?  I think I could have used it recently too.

Reply:
Fine by me.

Reply:
-> https://github.com/dlang/dmd/pull/8228